pkg/tbtc/heartbeat.go

```diff
@@ -111,16 +138,72 @@ func (ha *heartbeatAction) execute() error {
 	)
 	defer cancelHeartbeatCtx()
 
-	signature, _, err := ha.signingExecutor.sign(heartbeatCtx, messageToSign, ha.startBlock)
+	signature, activeOperatorsCount, signingEndBlock, err := ha.signingExecutor.sign(
+		heartbeatCtx,
+		messageToSign,
+		ha.startBlock,
+	)
+
+	// If there was no error and the number of active operators during signing
+	// was enough, we can consider the heartbeat procedure as successful.
+	if err == nil && activeOperatorsCount >= heartbeatSigningMinimumActiveOperators {
+		logger.Infof(
+			"successfully generated signature [%s] for heartbeat message [0x%x]",
+			signature,
+			ha.proposal.Message[:],
+		)
+
+		// Reset the counter for consecutive heartbeat failure.
+		*ha.failureCounter = 0
+
+		return nil
+	}
+
+	// If there was an error or the number of active operators during signing
+	// was not enough, we must consider the heartbeat procedure as a failure.
 	if err != nil {
-		return fmt.Errorf("cannot sign heartbeat message: [%v]", err)
+		logger.Infof("error while generating heartbeat signature: [%v]", err)
+	} else {
+		logger.Infof(
+			"not enough active operators during signing; required [%d]: "+
+				"actual [%d]",
+			activeOperatorsCount,
+			heartbeatSigningMinimumActiveOperators,
+		)
 	}
 
-	logger.Infof(
-		"generated signature [%s] for heartbeat message [0x%x]",
-		signature,
-		ha.proposal.Message[:],
+	// Increment the heartbeat failure counter.
+	*ha.failureCounter++
+
+	// If the number of consecutive heartbeat failures does not exceed the
+	// threshold do not notify about operator inactivity.
+	if *ha.failureCounter < heartbeatConsecutiveFailureThreshold {
+		logger.Infof(
+			"leaving without notifying about operator inactivity; current "+
+				"heartbeat failure count is [%d]",
+			*ha.failureCounter,
+		)
+		return nil
+	}
+
+	// The value of consecutive heartbeat failures exceeds the threshold.
+	// Proceed with operator inactivity notification.
+	err = ha.inactivityClaimExecutor.publishClaim(
```

Comment:
Do we know the duration of `publishClaim`? Asking because:
- `heartbeatProposalValidityBlocks` is `300` blocks so ~1h
- `heartbeatRequestTimeoutSafetyMarginBlocks` is `25` blocks so ~5m

That said, signing can take up to ~55m (see `heartbeatCtx`) and leave no time for `publishClaim`. I think we need to adjust the above parameters to include `publishClaim` in the action time.

Reply:
As discussed we decided to make the synchronization context-driven. See 9ad6371c9f14856f97900d95e627f05ed29dc344.